npe2 docs (#3769)
Napari/napari: docs/plugins/npe2_migration_guide.md

(npe2-migration-guide)=

> 
>     tutorial
>     - hookspec/contribution point mapping
>     - tooling - cli for generating a manifest
>     - where to use original plugin system
>         - what's unsupported at the moment

Author: ```suggestion
```

this case we're asking for help with the `convert` command:

```
Usage: npe2 convert [OPTIONS] PLUGIN_NAME

  Convert existing plugin to new manifest.

Arguments:
  PLUGIN_NAME  The name of the plugin to convert  [required]

Options:
  --format [yaml|json|toml]  [default: yaml]
  --out PATH
  --help                     Show this message and exit.
```

### 2. Install your `napari-plugin-engine` based plugin.

As an example, we'll walk through the process using the `napari-animation`
plugin.

We checkout the plugin and install it as a local editable package.

```bash
git clone https://github.com/napari/napari-animation.git
cd napari-animation
pip install -e .
```

### 3. Inspect package metadata

The `napari-animation` package defines it's core metadata in `setup.cfg`.
Inside, it defines an _entry point group_. That section should initially
contain:

```ini
[options.entry_points]
napari.plugin =
    animation = napari_animation
```

That metadata gives the name of the plugin: "animation". The name is used in
the next step. Later we're going to come back and update this section.

### 4. Generate the plugin manifest

> Tse the npe2 command in the terminal:

Reviewer: I don't know what `Tse` is supposed to be here but I suspect we need to fix this 😆

Author: ```suggestion
To create the manifest, use the npe2 command in the terminal:
```

Author: 😬

```bash
# we're in the napari-animation directory
npe2 convert animation --out napari_animation/napari.yaml
```

This generates `napari_animation/napari.yaml` with contents:

```yaml
name: napari_animation
contributions:
  commands:
    - id: napari_animation.AnimationWidget
      python_name: napari_animation._hookimpls:AnimationWidget
      title: Create Wizard
  widgets:
    - command: napari_animation.AnimationWidget
      name: Wizard
```

In this case, the manifest could be created without any intervention, but
sometimes the generated manifest needs to be edited. The conversion tool will
let you know when this happens.

### 5. Update the package metadata

Finally, we need to make sure `napari-animation` defines the npe2 entry point

> group. This lets npe2 where to find the plugin manifest file.

Reviewer: ```suggestion
group. This lets the plugin engine know where to find the plugin manifest file.
```

```ini
[options.entry_points]
# napari.plugin =
  # animation = napari_animation
napari.manifest =
    napari-animation = napari_animation:napari.yaml
```

Since we're interested in creating distributions of our package, make sure the
manifest gets included as package data:

```ini
[options]
include_package_data = True
   ...

[options.package_data]
napari_animation =
    napari.yaml
```

All done! Update the local package metadata by repeating:

```
> pip install -e .
```

and the next time napari is run `napari-animation` will be discovered as an
`npe2` plugin!

## Migration details

Existing `napari-plugin-engine` plugins expose functionality via _hook
implementations_. These are functions decorated to indicate they fullfil a
_hook specification_ described by napari. Though there are some exceptions,
most _hook implementations_ can be straightforwardly mapped as a
_contribution_ in the `npe2` manifest. More information can be found in the

> {ref}`npe2-manifest-spec`.

Reviewer: ```suggestion
{ref}`npe2-manifest-spec` - Look for "calling conventions" for each field.
```

`npe2` provides a command-line tool that will generate plugin manifests by
inspecting exposed _hook implementations_. Below, we will walk through the
kinds of migrations `npe2 convert` helps with.

For each type of _hook specification_ there is a corresponding section below.

> Each lists the _hook specifications_ that a relevant to that section and an

Reviewer: ```suggestion
Each lists the _hook specifications_ that are relevant to that section and an
```

In the example below, the svg writer accepts a set of layers with 0 or more
images, and 0 or more label layers, and so on. It will not accept surface
layers, so if any surface layer is present this writer won't be invoked.

Because layer type constraints are specified in the manifest, no plugin code
has to be imported or run until a compatible writer is found.

#### napari_hook_spec

```python
def napari_get_writer(
    path: str, layer_types: List[str]
) -> Optional[WriterFunction]
```

Where the `WriterFunction` is something like:

```python
def writer(
    path: str, layer_data: List[Tuple[Any, Dict, str]]
    ) -> List[str]
```

#### Example npe2 contribution

```yaml
name: napari_svg
display_name: napari SVG
entry_point: napari_svg
contributions:
  commands:
    - id: napari_svg.svg_writer
      title: Write SVG
      python_name: napari_svg.hook_implementations:writer
  writers:
    - command: napari_svg.svg_writer
      layer_types: ["image*", "labels*", "points*", "shapes*", "vectors*"]
      filename_extensions: [".svg"]
```

### Widgets

`napari_experimental_provide_dock_widget` hooks return another function that

> can be used to instance a widget and, optionally, arguments to be passed to

Reviewer: ```suggestion
can be used to instantiate a widget and, optionally, arguments to be passed to
```

that function.

In contrast the callable for an `npe2` widget contribution is bound to the

> function actually instancing the widget. It accepts only one argument: a

Reviewer: ```suggestion
function actually instantiates the widget. It accepts only one argument: a
```

Author: ```suggestion
function actually instantiating the widget. It accepts only one argument: a
```
